### PR TITLE
Clean up utility functions.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR removes unused utility functions and cleans up test coverage of the `utility.py` module.

Notes:
- The `_query_yes_no` function doesn't need test coverage (it's strictly for interactive use on the terminal).
- The `_split_and_print_progress` function can probably be simplified and replaced with tqdm, if it is needed at all. The function is used only once, to print info-level logging for updates to the in-memory state point cache. This should be investigated further, but I've made the function private for now.
- There was an important edge case of `_nested_dicts_to_dotted_keys` that wasn't being tested, when a value is an empty dict. I added a test that covers that edge case.

## Motivation and Context
Cleanup for signac 2.0.

## Types of Changes
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

File: tests/test_project.py [hunk 1]
```diff
@@ -805,8 +805,9 @@ def test_schema(self):
         for i in range(10):
             self.project.open_job(
                 {
-                    "const": 0,
+                    "const1": 0,
                     "const2": {"const3": 0},
+                    "const4": {},
```
Comment (bdice) [inline tests/test_project.py:810]: This edge case was being missed until I added this test: https://github.com/glotzerlab/signac/blob/257c93169709039ef4ca54e8359aace31c62b0c8/signac/contrib/utility.py#L339-L340

File: signac/contrib/project.py [hunk 2]
```diff
@@ -41,7 +41,7 @@
 from .indexing import _SignacProjectCrawler
 from .job import Job
 from .schema import ProjectSchema
-from .utility import _mkdir_p, _nested_dicts_to_dotted_keys, split_and_print_progress
+from .utility import _mkdir_p, _nested_dicts_to_dotted_keys, _split_and_print_progress
```
Comment (csadorf) [inline signac/contrib/project.py:44]: If I've had to guess the function was introduced prior to having the tqdm dependency or because there was some kind of threading issue or so with tqdm. But I agree, we should certainly try to replace it with tqdm if possible.

**Review by csadorf — APPROVED** [review]
